### ISSUE TITLE
JRuby/S3 Bug: Failed to open TCP connection

### ISSUE THREAD

**Issue by aryeh-looker** [aws/aws-sdk-ruby#1843]
### Issue description

Under JRuby, the SDK produces an invalid URI when attempting to upload a file to S3 when an incorrect region is specified. Expected behavior is that the multipart upload will succeed. 

### Gem name

`gem 'aws-sdk-s3', '1.8.0'`

### Version of Ruby, OS environment

`jruby-9.17.0.0` on Ubuntu and Mac OS X

### Code snippets / steps to reproduce

* save the script below to a file
* use `jruby-9.17.0.0` (i.e. install `rvm`, download/switch to to JRuby)
* run the script below

Note that this script works under Ruby, but fails under JRuby.

```ruby
require 'securerandom'
require 'aws-sdk-s3'

#########################
# Fill in these values before running this script.
# The `BUCKET` and `REGION` must mismatch for the repro. to be successful
ACCESS_KEY_ID = ''
SECRET_ACCESS_KEY = ''
BUCKET = ''
REGION = ''
#########################

def upload_things(region:, bucket:)
    key = SecureRandom.hex
    client = Aws::S3::Client.new(region: region,
                                 access_key_id: ACCESS_KEY_ID,
                                 secret_access_key: SECRET_ACCESS_KEY)

    # This call will succeed. Under the hood, the SDK makes two HTTP requests:
    # 1) one to the API specifying the wrong region; it will reply with XML that describes the correct region in Aws::S3::Plugins::S3Signer
    # 2) one to the API with the correct region
    multipart_upload = client.create_multipart_upload(bucket: bucket,
                                                      key: key)

    parts = []
    3.times do |part_number|
		puts "Starting part #{part_number + 1} (#{bucket})"
                 
                # This call will fail and raise an exception because the URI is invalid.
                # This is because the call above will have cached the region for the bucket.
                # `Aws::S3::Plugins::S3Signer` ill attempt to fetch the region from cache
                # and it will be corrupted. The corrupted value will be part of the URI, which
                # will be invalid. 
		part = client.upload_part(body: "#{part_number + 1}\n" * 5e6, # Overcome minimum size limitation
		                   	      bucket: bucket,
		                   	      key: key,
		                   	      part_number: part_number + 1,
		                   	      upload_id: multipart_upload.upload_id)
		parts << {etag: part['etag'], part_number: part_number + 1}
    end

    completion = client.complete_multipart_upload(bucket: bucket,
                                                  key: key,
                                                  multipart_upload: { parts: parts },
                                                  upload_id: multipart_upload.upload_id)

    [key, client, multipart_upload, parts, completion]
end

def run
  begin
    # The region and bucket should be mismatched
    result = upload_things(region: REGION, bucket: BUCKET)
  rescue Exception => e
    raise e
  end
end

run()
```

---

This is a repro. for the related issue https://github.com/aws/aws-sdk-ruby/issues/1785.

**Comment by aryeh-looker** [comment]
Little more background — in our case, `bucket`/`region` pairs are user-provided (and we have many users; not a simple case of server misconfiguration, etc.). It will likely be possible for us to work around this bug by using the `#get_bucket_location` API before making calls to upload files to S3.

**Comment by aryeh-looker** [comment]
Looks like this is due to a bug in JRuby that is slated to be fixed in the `9.1.18.0` release https://github.com/jruby/jruby/issues/5203#issuecomment-408656779.
